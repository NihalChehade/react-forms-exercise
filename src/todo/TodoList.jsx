import React, {useState} from 'react'
import {v4 as UUID} from 'uuid';
import NewTodoForm from './NewTodoForm'
import Todo from './Todo'
const TodoList = () => {
  const [todos, setTodos] = useState([]);

  const addTodo = (todo) =>{
    setTodos([...todos, {key: UUID(), value :todo}]);
  }

  const removeTodo = (key)=>{
    setTodos(todos.filter((obj)=>obj.key!=key))
  }
  return (
    <div>
      <NewTodoForm addTodo={addTodo} />
      <ol>
      {todos.map(todo=><Todo id={todo.key} key={todo.key} value={todo.value} removeTodo={removeTodo}/>)}
      </ol>
    </div>
  )
}

export default TodoList