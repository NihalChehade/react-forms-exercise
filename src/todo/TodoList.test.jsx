import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import TodoList from './TodoList';

// Smoke Test
test('renders TodoList component without crashing', () => {
  render(<TodoList />);
});

// Snapshot Test
test('matches snapshot', () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

// Critical Business Logic Tests
test('adds a new todo when addTodo is called', () => {
  render(<TodoList />);
  
  // Initially, no todos should be present
  expect(screen.queryAllByRole('listitem')).toHaveLength(0);
  
  // Add a new todo
  fireEvent.click(screen.getByText('Add Todo!'));

  // Now there should be one todo
  expect(screen.queryAllByRole('listitem')).toHaveLength(1);
});

test('removes a todo when removeTodo is called', () => {
  render(<TodoList />);
  
  // Add a new todo
  fireEvent.click(screen.getByText('Add Todo!'));

  // Remove the todo
  fireEvent.click(screen.getByText('X'));
  
  // No todos should be present now
  expect(screen.queryAllByRole('listitem')).toHaveLength(0);
});
