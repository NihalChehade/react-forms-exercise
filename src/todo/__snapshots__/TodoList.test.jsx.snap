// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <label
        for="todo"
      >
        Todo:
      </label>
      <input
        id="todo"
        name="todo"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo!
      </button>
    </form>
    <ol />
  </div>
</DocumentFragment>
`;
